apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.IntegrationTests/SimplePubSubTest.cs

```diff
@@ -73,29 +79,50 @@ public SimplePubSubTest(PubsubFixture fixture)
                     }
                 )).ConfigureAwait(false);
             var simpleSubscriber = await SimpleSubscriber.CreateAsync(subscriptionName,
-                settings: timeouts == null ? null : new SimpleSubscriber.Settings
+                settings: new SimpleSubscriber.Settings
                 {
-                    StreamAckDeadline = timeouts.Value
+                    StreamAckDeadline = timeouts,
+                    FlowControlSettings = new FlowControlSettings(maxMessagesInFlight, null)
                 }).ConfigureAwait(false);
 
             Console.WriteLine("Topic, Subscription, SimplePublisher and SimpleSubscriber all created");
 
             // Subscribe
             object recvLock = new object();
             int recvCount = 0; // Count of received messages
+            int dupCount = 0; // Count of duplicate messages
             long recvSum = 0L; // Sum of bytes of received messages
+            var recvedIds = new ConcurrentDictionary<int, bool>();
+            var nackedIds = new HashSet<int>();
             Task subTask = simpleSubscriber.StartAsync((msg, ct) =>
             {
-                var localRecvCount = Interlocked.Increment(ref recvCount);
-                if (localRecvCount <= initialNackCount)
+                int id = BitConverter.ToInt32(msg.Data.ToArray(), 0);
+                lock (nackedIds)
                 {
-                    return Task.FromResult(SimpleSubscriber.Reply.Nack);
+                    if (nackedIds.Count < initialNackCount)
+                    {
+                        if (nackedIds.Add(id))
+                        {
+                            // This ID not already nacked
+                            var localRecvCount = Interlocked.Increment(ref recvCount);
+                            return Task.FromResult(SimpleSubscriber.Reply.Nack);
+                        }
+                    }
+                }
+                bool wasAdded = recvedIds.TryAdd(id, false);
+                if (wasAdded)
+                {
+                    var localRecvCount = Interlocked.Increment(ref recvCount);
+                    Interlocked.Add(ref recvSum, msg.Data.Sum(x => (long)x));
+                    if (localRecvCount == cancelAfterRecvCount || localRecvCount >= messageCount + initialNackCount)
```

Comment:
Why would it be greater than messageCount + initialNackCount?

Reply:
It shouldn't be, but this is just in case...